keps/sig-instrumentation/3077-contextual-logging/README.md

```diff
@@ -0,0 +1,1128 @@
+the logger once and then use that instance.
+
+More expensive than logger lookup are `WithName` and `WithValues` and creating
+a new context with the modified logger. This needs to be used cautiously in
+performance-sensitive code. A possible compromise is to enhance logging with
+such additional information only at higher log levels.
+
+#### Transition
+
+Once some components start to rely on `k8s.io/klogr` exclusively instead of
+klog, users of those components must remember to initialize klog *and* the
+default logger in `k8s.io/klogr`. Features provided only by klog like writing
+to files become unavailable.
+
+This breaking change can be postponed until after the [deprecation of klog
+features](https://github.com/kubernetes/enhancements/issues/2845) is complete
+by using a logger which passes log messages through to klog. That allows
+converting components already now. Then later, the klog dependency in
+`k8s.io/klogr` can be replaced with a stand-alone logger which produces the same
+text output.
+
+#### Pitfalls during usage
+
+Code reviews must catch the following incorrect patterns.
+
+##### Logger in context and function not consistent
+
+Incorrect:
+
+```go
+func foo(ctx context.Context) {
+   logger := klogr.FromContext(ctx)
+   ctx = klogr.NewContext(ctx, logger.WithName("foo"))
+   doSomething(ctx)
+   // BUG: does not include WithName("foo")
+   logger.Info("Done")
+}
+```
+
+Correct:
+
+```go
+func foo(ctx context.Context) {
+   logger := klogr.FromContext(ctx).WithName("foo")
+   ctx = klogr.NewContext(ctx, logger)
+   doSomething(ctx)
+   logger.Info("Done")
+}
+```
+
+In general, manipulating a logger and the corresponding context should be done
+in separate lines.
+
+##### Overwriting context not needed initially, forgotten later
+
+Initial, correct code with contextual logging:
+
+```go
+func foo(ctx context.Context) {
+   logger := klogr.FromContext(ctx).WithName("foo")
+   doSomething(logger)
+   logger.Info("Done")
+}
+```
+
+A line with `ctx = klogr.NewContext(ctx, logger)` could be added above (it
+compiles), but it causes unnecessary overhead and some linters complain about
+"new value not used".
+
+However, care then must be taken when adding code later which uses `ctx`:
+
+```go
+func foo(ctx context.Context) {
+   logger := klogr.FromContext(ctx).WithName("foo")
+   doSomething(logger)
+   // BUG: ctx does not contain the modified logger
+   doSomethingWithContext(ctx)
+   logger.Info("Done")
+}
+```
+
+##### Redundant key/value pairs
+
+When the caller already adds a certain key/value pair to the logger, the callee
+should still add it as parameter in log calls where it is important. That is
+because the contextual logging feature might be disabled, in which case adding
+the value in the caller is a no-op. Another reason is that it is not always
+obvious whether a value is part of the logger or not.
+
+If there are duplicate keys, the text output will only print the value from the
```

Comment:
This is a klog implementation detail, if so.  funcr will print 2 keys (WithValues() is pre-rendered for perf).

Reply:
True, but this is about the text output through `k8s.io/klogr` and that does deduplicate.